README

Just what is this and why did you write it?

Well, I wrote MySQL::Easy (bad name, I know) because I didn't like all the
work I had to do maintaining and working with the DBI methods to keep my
mysql connection alive.  I found I had to do the same simple lines of code
over and over and I didn't like it.  That is all.

Examples:

my $dbh = DBI->connect("DBI::mysql:testdb:host=something:port=something;opt;opt;opt");
my $dbo = new MySQL::Easy("testdb");

Easy reads your $ENV{HOME}/.my.cnf for unp when you don't specify the
username/pass directly.  Handy?!?!  Yes.

my $sth1 = $dbh->prepare("select something from something where something=something");
my $sth2 = $dbo->ready("select something from something where something=something");

$sth1 looks pretty much the same as $sth2... but $sth2 will check to see if
the connection is still there and if it isn't, will rebuild the connection
and re-prepare the sth automatically -- ie, no errors.

Other than these two little things, this package is probably not necessary.
However I'm aware of literally dozens that use it -- including myself -- so
I continue to maintain it.

- Paul


Changes
2.1020: Wed Nov 19 2014
   - need more of this easy::st logic to recover from these stupid connection crashes
   - I'm still having all sorts of connection problems on that one host, and I found
     several places where connections drop inexplicably.  building tests.
   - I also increased the number of reconnect retries and the wait time between the retries
     (server outages tend to be bursty … unless the program should simply crash, it's probably
      better to just wait until it's back.)

2.1019: Sun Nov 16 2014
   - try to repair connections on dhb methods like selectall_arrayref

2.1017: Sat Feb 08 2014
   - make the autoloads actually autoload

2.1015: Thu Jan 30 2014
   - improved autoloader for dbh

2.1014: Fri Oct 04 2013
   - firstrow()

2.1012: Sun Nov 25 2012
   - allow for DBIx::ContextualFetch, etc

2.1011: Tue Oct 04 2011
   - disable spurious warning

2.1010: Thu Jun 30 2011
   - more fixes on error checking

2.1009: Wed Jun 29 2011
   - slightly better error handling

2.1008: Tue May 18 2010
   - alternate locations for username and password info
   - documented the new locations

2.1007: Tue May 11 2010
   - more introduced bugs

2.1006: Mon May 10 2010
   - I made various improvements to the API, nothing worth
     mentioning.  Just little things to the new() call -- see the
     POD.  Hrm, $dbo->firstval() is worth noting -- see the POD.

2.1003: Fri Mar 26 2010
   - got utf8 working

2.1002: Tue Aug 04 2009
   - make sure autoraise is explicit
   - remove or die $dbo->errstr from docs

2.1001: Wed May  6 07:22:23 EDT 2009
   - the only place the tests actually run (my house)
     ... they weren't being run.

2.1000: Wed May  6 07:22:23 EDT 2009
   - I cleaned up the pod a lot
   - modernized the makefile
   - fixed various mundane things
   - got rid of version.pm (cursed object)

2.0.4: Mon Jun  9 20:55:50 EDT 2008
   - added a smarter last_insert_id

2.0.3: Tue May 27 12:04:27 EDT 2008
   - (wow, use version has to be on the same line as our $VERSION
      for older EUMM)

2.0.2: Mon May 26 15:56:50 EDT 2008
   - I might as well do a new version of everything using
     version.pm while I wait for all my CPAN deletes to go
     through.  I just discovered that version.pm doesn't/didn't
     go core until perl 5.9...  I really wish I'd have prereqed
     it everywhere I used it, but I didn't.  *sigh*

2.0.1: Sat May 24 09:37:56 EDT 2008
   - using version.pm as suggested by Andreas Koenig

2.0.0: May 29th 2007
   - I have re moved this package back into the MySQL::Easy
     namespace.  It's a terrible name, but the whole package is
     very mysql specific and the DBIx::Easy::MySQL namespace did
     not work out well at all.

   - I also stripped out the sqlite portions.  There was nothing
     to make easy about using it.

1.42:  Tue Feb 20 11:29:41 EST 2007
   - jettero made this still test ok even if you don't have
     DBD::SQLite

1.41:  Fri May 19 21:27:43 EDT 2006
   - jettero added light_lock

1.40:  Thu Mar 30 07:25:45 EST 2006
   - jettero added sqlite

1.33:  Thu Jul 14 08:26:49 EDT 2005
   - jettero fixed the MANIFEST to include the new tests
     (not that they run at all without being on my devel
     machine).

1.32:  Wed Jul 13 14:11:35 EDT 2005
   - jettero added a test for the selectall_arrayref() ::sth
     problem

1.31:  Wed Jul 13 14:10:34 EDT 2005
   - jettero added error checking to the Easy AUTOLOAD to
     prevent passing the AUTOLOADing Easy version of sth to
     DBI. It checks for SvMAGICAL and Jet can fake it
     without doing a proper @ISA=qw(DBI::st). The
     disadvantage of that redesign ... is a redesign.

1.30:  Wed May 25 17:12:15 EDT 2005
   - jettero created a comprehensive test to cover the new
     stupid bugs...

1.29:  Wed May 25 16:52:56 EDT 2005
   - jettero fixed an AUTOLOAD bug in the new ::sth module.
     Dumb.

1.28:  Wed May 25 15:41:17 EDT 2005
   - jettero fixed a couple STUPID bugs in the tests.

1.27:  Wed May 25 15:17:48 EDT 2005
   - jettero made the :sth returned by ready() a special
     MySQL::Easy version that auto-reloads itself in the
     case of the DBI connection fading and re-connecting.
     This way you don't have to re-prepare the same
     statements over and over -- by hand anyway.

1.26:  Fri Feb 11 07:25:31 EST 2005
   - jettero added a check_warnings function

1.22:  Fri Oct  3 14:12:42 EDT 2003
   - jettero made a minor change to the sub handle {} that
     now pings the dbh to decide if it's still there.

1.21:  Mon May 19 14:23:29 EDT 2003
   - jettero added a better last_insert function

1.2:  Tue Oct 29 09:13:58 EST 2002
   - jettero added the johsr bind_execute function

1.19:  Tue Oct 29 08:08:03 EST 2002
   - jettero fixed some docs typos and added Josh's existing
     handle support.

1.18:  Thu Oct  3 09:40:15 EDT 2002
   - jettero fixed a really stupid syntax bug ...

1.17:  Fri Aug 30 07:01:19 EDT 2002
   - jettero fixed two very stupid bugs in the AUTOLOADer
     ... lost my webserver at 5am due to one of them. :(

1.16:  Wed Aug 28 17:04:27 EDT 2002
   - jettero made a heafty change .... and introduced that
     AUTOLOAD ... why didn't I think of that before!?!?

1.15:  Thu Jun  6 22:48:47 EDT 2002
   - jettero did some things

1.14:  Tue Jun  4 20:01:36 EDT 2002
   - jettero fixed the test.t

1.13:  Tue Jun  4 19:59:07 EDT 2002
   - jettero made some distribution cleanups

1.12:  Tue Jun  4 19:57:17 EDT 2002
   - jettero made a couple changes ...



Easy.pm

package MySQL::Easy::sth;

use Carp;
use common::sense;

our $AUTOLOAD;
our $RESTARTABLE_ERRORS       = qr/(?:has gone away|Lost connection)/;
our $SLEEP_ON_RECONNECT_RETRY = 1;
our $RECONNECT_RETRIES        = 5;

# new {{{
sub new {
    my ($class, $mysql_e, $statement) = @_;
    my $this  = bless { s=>$statement, dbo=>$mysql_e }, $class;

    $this->{sth} = $this->{dbo}->handle->prepare( $statement );

    return $this;
}
# }}}
# bind_execute {{{
sub bind_execute {
    my $this = shift;

    $this->{sth}->execute;
    $this->{sth}->bind_columns( @_ );

    return $this;
}
# }}}
# {{{ sub repair_statement
sub repair_statement {
    my $this = shift;

    $this->{sth} = $this->{dbo}->handle->prepare( $this->{s} );
    $this->{repair} ++;

    return $this;
}

# }}}
# AUTOLOAD {{{
sub AUTOLOAD {
    my $_self = shift;
    my $sub = $AUTOLOAD;
       $sub = $1 if $sub =~ m/::(\w+)$/;

    return unless $_self->{sth}; # I should be dead?
    croak "$sub is not a member of " . ref($_self->{sth}) unless $_self->{sth}->can($sub);

    *{ __PACKAGE__ . "::$sub" } = sub {
        my $this = shift;
        my $tries = $RECONNECT_RETRIES;

        return unless $this->{sth}; # I should be dead?

        my $wa = wantarray;
        my ($err, $warn, $ret, @ret);

        # warn "DEBUG: FYI, $$-$this is loading $sub()";

        EVAL_IT: eval {
            no strict 'refs';
            local $SIG{__WARN__} = sub { $warn = "@_"; };

            if( $wa ) {
                @ret = $this->{sth}->$sub( @_ );

            } else {
                $ret = $this->{sth}->$sub( @_ );
            }
        };

        $err = $@;

        if( $warn and not $err ) {
            $err = $warn;
            chomp $err;
        }

        if( $err ) {
          # my @c = caller;
          # my $p = "at $c[1] line $c[2], prepared at $this->{_ready_caller}[1] line $this->{_ready_caller}[2]\n";
            my $p = "(prepared at $this->{_ready_caller}[1] line $this->{_ready_caller}[2])";

            1 while $err =~ s/\s+at(?:\s+\S+)?\s+line\s+\d+\.?$//;

            # ERROR executing execute(): DBD::mysql::st execute failed: You have an error in your SQL syntax; check the manual
            $err =~ s/DBD::mysql::sth? execute failed:\s*//;

            if( $err =~ $RESTARTABLE_ERRORS ) {
                if( $sub eq "execute" ) {
                    $this->repair_statement;
                    $warn = undef;

                    sleep $SLEEP_ON_RECONNECT_RETRY if $SLEEP_ON_RECONNECT_RETRY > 0;
                    goto EVAL_IT if ((--$tries) >= 0);
                    $err .= " ($tries/$RECONNECT_RETRIES)";

                } else {
                    croak "MySQL::Easy::sth can only recover from connection problems during execute(): $err $p";
                }
            }

            croak "ERROR executing $sub(): $err $p";
        }

        return ($wa ? @ret : $ret);
    };

    return $_self->$sub(@_);
}
# }}}
# DESTROY {{{
sub DESTROY {
    my $this = shift;

    # warn "MySQL::Easy::sth is dying"; # This is here to make sure we don't normally die during global destruction.
                                        # Once it appeared to function correctly, it was removed.
                                        # Lastly, we would die during global dest iff: our circular ref from new() were not removed.
                                        # Although, to be truely circular, the MySQL::Easy would need to point to this ::sth also
                                        # and it probably doesn't.  So, is this delete paranoid?  Yeah...  meh.
    delete $this->{dbo};
}
# }}}

package MySQL::Easy;

use Carp ();
use common::sense;
use Scalar::Util qw(blessed);
use overload fallback=>1, '""' => sub { ref($_[0]) . "($_[0]{dbase})" };

use DBI;

our $AUTOLOAD;
our $VERSION = "2.1020";
our $CNF_ENV = "ME_CNF";
our $USER_ENV = "ME_USER";
our $PASS_ENV = "ME_PASS";
our $HOME_ENV = "HOME";
our @MY_CNF_LOCATIONS = (
    $ENV{$CNF_ENV}, "$ENV{$HOME_ENV}/.my.cnf", "/etc/mysql-easy.cnf", "/etc/mysql/my.cnf"
);

# {{{ sub mycroak
sub mycroak(;$) {
    my $error = shift;

    my $i = 1;
    my @c = caller(  $i);
       @c = caller(++$i) while $c[0] eq __PACKAGE__;

    chomp $error;

    1 while
    $error =~ s{\s+at\s+\S+\s+line\s+\d+\.}{}g;
    $error =~ s{\s+\(prepared at $c[1] line $c[2]\)}{}; # this would be a dup error in this case

    #arn "<<<$error\[$i]>>>";

    Carp::croak($error);
}

# }}}

# AUTOLOAD {{{
sub AUTOLOAD {
    my $_self = shift;
    my $sub   = $AUTOLOAD;
       $sub   = $1 if $sub =~ m/::(\w+)$/;

    {
        my $handle = $_self->handle;
        mycroak "$sub is not a member of " . ref($handle) unless $handle->can($sub);
    }

    *{ __PACKAGE__ . "::$sub" } = sub {
        my $this = shift;
        my $handle = $this->handle;

        my $wa = wantarray;
        my ($err, $warn, $ret, @ret);
        my @oargs = @_;
        my $tries = $RECONNECT_RETRIES;

        EVAL_IT: my $eval_result = eval {
            $warn = undef;
            local $SIG{__WARN__} = sub { $warn = "@_"; };

            my @cargs = @oargs;
            $cargs[0] = $cargs[0]->{sth} if blessed $cargs[0] and exists $cargs[0]->{sth};

            if( wantarray ) {
                @ret = $handle->$sub(@cargs);

            } else {
                $ret = $handle->$sub(@cargs);
            }

            !$warn
        };

        unless( $eval_result ) {
            $err = $warn;

            if( $err =~ $RESTARTABLE_ERRORS ) {
                if( blessed $oargs[0] ) {
                    if( $oargs[0]->isa("MySQL::Easy::sth") ) {
                        $oargs[0]->repair_statement;
                    }
                }

                sleep $SLEEP_ON_RECONNECT_RETRY if $SLEEP_ON_RECONNECT_RETRY > 0;
                goto EVAL_IT if ((--$tries) >= 0);
                $err .= " ($tries/$RECONNECT_RETRIES)";
            }

            if( $warn and not $err ) {
                $err = $warn;
                chomp $err;
            }

            $err =~ s/DBD::mysql::dbh? \S+ failed:\s*//;
            mycroak "ERROR executing $sub(): $err";
        }

        return ($wa ? @ret : $ret);
    };

    #arn "created method $sub, calling";

    return $_self->$sub(@_);
}
# }}}

# check_warnings {{{
sub check_warnings {
    my $this = shift;
    my $sth  = $this->ready("show warnings");

    # mysql> show warnings;
    # +---------+------+------------------------------------------+
    # | Level   | Code | Message                                  |
    # +---------+------+------------------------------------------+
    # | Warning | 1265 | Data truncated for column 'var' at row 1 |
    # +---------+------+------------------------------------------+

    my @warnings;

    execute $sth or die $this->errstr;
    while( my $a = fetchrow_arrayref $sth ) {
        push @warnings, $a;
    }
    finish $sth;

    if( @warnings ) {
        $@ = join("\n", map("$_->[0]($_->[1]): $_->[2]", @warnings)) . "\n";

        return 0;
    }

    return 1;
}
# }}}
# new {{{
sub new {
    my $this = shift;

    $this = bless {}, $this;

    $this->{dbase} = shift; mycroak "dbase = '$this->{dbase}'?" unless $this->{dbase};
    $this->{dbh} = $this->{dbase} if ref($this->{dbase}) and $this->{dbase}->isa("DBI::db");

    my $args = shift;
    my $tr   = ref($args) ? delete $args->{trace} : $args;

    $this->trace($tr) if $tr and $this->{dbh};

    if( ref $args ) {
        for my $k (keys %$args) {
            my $f;

            if( $this->can($f = "set_$k") ) {
                $this->$f($args->{$k});

            } else {
                mycroak "unrecognized attribute: $k"
            }
        }
    }

    return $this;
}
# }}}
# do {{{
sub do {
    my $this = shift; return unless @_;
    my $sql  = shift;
    my $r; eval { $r = $this->ready($sql)->execute(@_); 1 } or mycroak $@;
    return $r;
}
# }}}
# light_lock {{{
sub light_lock {
    my $this   = shift; return unless @_;
    my $tolock = join(", ", map("$_ read", @_));

    $this->do("lock tables $tolock");
}
# }}}
# lock {{{
sub lock {
    my $this   = shift; return unless @_;
    my $tolock = join(", ", map("$_ write", @_));

    $this->do("lock tables $tolock");
}
# }}}
# unlock {{{
sub unlock {
    my $this = shift;

    $this->do("unlock tables");
}
# }}}
# ready {{{
sub ready {
    my $this = shift;

    my $i = 0;
    my $sth = MySQL::Easy::sth->new( $this, @_ );
       $sth->{_ready_caller} = [ caller(  $i) ];
       $sth->{_ready_caller} = [ caller(++$i) ] while $sth->{_ready_caller}[0] eq __PACKAGE__;

    return $sth;
}
# }}}
# firstcol {{{
sub firstcol {
    my $this = shift;
    my $query = shift;

    my $r;
    eval { $r = $this->selectcol_arrayref($query, undef, @_); 1 }
        or mycroak $@;

    return wantarray ? @$r : $r;
}
# }}}
# firstval {{{
sub firstval {
    my $this = shift;
    my $query = shift;

    my $r;
    eval { $r = $this->selectcol_arrayref($query, undef, @_); 1 }
        or mycroak $@;

    return $r->[0];
}
# }}}
# firstrow {{{
sub firstrow {
    my $this = shift;
    my $query = shift;

    my $r;
    eval { $r = $this->selectrow_arrayref($query, undef, @_); 1 }
        or mycroak $@;

    return wantarray ? @$r : $r;
}
# }}}
# thread_id {{{
sub thread_id {
    my $this = shift;

    return $this->handle->{mysql_thread_id};
}
# }}}
# last_insert_id {{{
sub last_insert_id {
    my $this = shift;

    # return $this->firstcol("select last_insert_id()")->[0];
    # return $this->handle->{mysql_insertid};
    return $this->handle->last_insert_id(undef,undef,undef,undef);
}
# }}}
# DESTROY {{{
sub DESTROY {
    my $this = shift;

    $this->{dbh}->disconnect if $this->{dbh};
}
# }}}
# handle {{{
sub handle {
    my $this = shift;

    return $this->{dbh} if defined($this->{dbh}) and $this->{dbh}->ping;
    # warn "WARNING: MySQL::Easy is trying to reconnect (if possible)" if defined $this->{dbh};

    ($this->{user}, $this->{pass}) = $this->unp unless $this->{user} and $this->{pass};

    $this->{host}  = "localhost" unless $this->{host};
    $this->{port}  =      "3306" unless $this->{port};
    $this->{dbase} =      "test" unless $this->{dbase};
    $this->{trace} =           0 unless $this->{trace};

    if( $this->{dbh} ) {
        eval {
            local $SIG{__WARN__} = sub {};  # Curiously, sometimes we do have a handle, but the ping doesn't work.
                                            # If we replace the handle, DBI complains about not disconnecting.
                                            # If we disconnect, it complains about not desting statement handles.
                                            # Heh.  It's gone dude, let it go.
            $this->{dbh}->disconnect;
        };
    }

    $this->{dbh} =
    DBI->connect("DBI:mysql:$this->{dbase}:host=$this->{host}:port=$this->{port}",
        $this->{user}, $this->{pass}, {

            RaiseError => ($this->{raise} ? 1:0),
            PrintError => ($this->{raise} ? 0:1),

            AutoCommit => 0,

            mysql_enable_utf8    => 1,
            mysql_compression    => 1,
            mysql_ssl            => 1,
            mysql_auto_reconnect => 1,

        });

    mycroak "failed to generate connection: " . DBI->errstr unless $this->{dbh};

    $this->{dbh}->trace($this->{trace});

    return $this->{dbh};
}
# }}}
# unp {{{
sub unp {
    my $this = shift;

    return ($ENV{$USER_ENV}, $ENV{$PASS_ENV}) if $ENV{$USER_ENV} and $ENV{$PASS_ENV};

    my ($user, $pass, $file, $fh);

    for $file (@MY_CNF_LOCATIONS) {
        next unless -f $file;
        next unless open $fh, $file;

        while(<$fh>) {
            $user = $1 if m/user\s*=\s*(.+)/;
            $pass = $1 if m/password\s*=\s*(.+)/;

            return ($user, $pass) if $user and $pass;
        }
    }

    die "unable to locate a username and password\n";
    return;
}
# }}}
# set_host set_user set_pass {{{
sub set_host {
    my $this = shift;

    $this->{host} = shift;
}

sub set_port {
    my $this = shift;

    $this->{port} = shift;
}

sub set_user {
    my $this = shift;

    $this->{user} = shift;
}

sub set_pass {
    my $this = shift;

    $this->{pass} = shift;
}

sub set_raise {
    my $this = shift;

    $this->{raise} = shift;
}
# }}}
# bind_execute {{{
sub bind_execute {
    my $this = shift;
    my $sql  = shift;

    my $sth = $this->ready($sql);

    $sth->execute            or return;
    $sth->bind_columns( @_ ) or return;

    return $sth;
}
# }}}

1;


MySQL.pm
# $Id: MySQL.pm,v 1.8 2006/05/20 01:27:45 jettero Exp $
# vi:fdm=marker fdl=0:

package MySQL::Easy::sth;

use strict;
use warnings;
use Carp;
use AutoLoader;

our $AUTOLOAD;

# new {{{
sub new {
    my ($class, $mysql_e, $statement) = @_;
    my $this  = bless { s=>$statement, dbo=>$mysql_e }, $class;

    $this->{sth} = $this->{dbo}->handle->prepare( $statement );

    return $this;
}
# }}}
# AUTOLOAD {{{
sub AUTOLOAD {
    my $this = shift;
    my $sub  = $AUTOLOAD;
    my $wa   = wantarray;

    return undef unless $this->{sth};
    # croak "this sth is defunct.  please don't call things on it." unless $this->{sth};

    $sub = $1 if $sub =~ m/::(\w+)$/;

    my $tries = 2;
    if( $this->{sth}->can($sub) ) {
        my @ret;
        my $ret;
        my $warn;

        # warn "DEBUG: FYI, $$-$this is loading $sub()";

        EVAL_IT: eval q/ 
            no strict 'refs';
            local $SIG{__WARN__} = sub { $warn = "@_"; };

            if( $wa ) {
                @ret = $this->{sth}->$sub( @_ );

            } else {
                $ret = $this->{sth}->$sub( @_ );
            }
        /;
         
        if( $warn and not $@ ) {
            $@ = $warn;
            chomp $@;
        }

        if( $@ ) {
            if( $@ =~ m/MySQL server has gone away/ ) {
                if( $sub eq "execute" ) {
                    $this->{sth} = $this->{dbo}->handle->prepare( $this->{s} );
                    $warn = undef;

                    goto EVAL_IT if ((--$tries) > 0);

                } else {
                    croak "MySQL::Easy::sth can only recover during execute(), $@";
                }
            }

            croak "ERROR executing $sub(): $@";
        }

        return ($wa ? @ret : $ret);

    } else {
        croak "$sub is not a member of " . ref($this->{sth});
    }
}
# }}}
# DESTROY {{{
sub DESTROY {
    my $this = shift;

    # warn "MySQL::Easy::sth is dying"; # This is here to make sure we don't normally die during global destruction.
                                        # Once it appeared to function correctly, it was removed.
                                        # Lastly, we would die during global dest iff: our circular ref from new() were not removed.
                                        # Although, to be truely circular, the MySQL::Easy would need to point to this ::sth also
                                        # and it probably doesn't.  So, is this delete paranoid?  Yeah...  meh.
    delete $this->{dbo};
}
# }}}

package MySQL::Easy;

use strict;
use warnings;
use Carp;
use AutoLoader;

use DBI;

our $AUTOLOAD;

1;

# AUTOLOAD {{{
sub AUTOLOAD {
    my $this = shift;
    my $sub  = $AUTOLOAD;

    $sub = $1 if $sub =~ m/::(\w+)$/;

    my $handle = $this->handle;

    if( $handle->can($sub) ) {
        no strict 'refs';
        return $handle->$sub( 
            (ref($_[0]) eq "MySQL::Easy::sth" ? $_[0]->{sth} : $_[0]), # cheap and not "gone away" recoverable
            @_[1 .. $#_],
        );

    } else {
        croak "$sub is not a member of " . ref($handle);
    }
}
# }}}

# check_warnings {{{
sub check_warnings {
    my $this = shift;
    my $sth  = $this->ready("show warnings");

    # mysql> show warnings;
    # +---------+------+------------------------------------------+
    # | Level   | Code | Message                                  |
    # +---------+------+------------------------------------------+
    # | Warning | 1265 | Data truncated for column 'var' at row 1 |
    # +---------+------+------------------------------------------+

    my @warnings;

    execute $sth or die $this->errstr;
    while( my $a = fetchrow_arrayref $sth ) {
        push @warnings, $a;
    }
    finish $sth;

    if( @warnings ) {
        $@ = join("\n", map("$_->[0]($_->[1]): $_->[2]", @warnings)) . "\n";

        return 0;
    }

    return 1;
}
# }}}
# new {{{
sub new { 
    my $this  = shift;

    $this = bless {}, $this;

    $this->{dbase} = shift; croak "dbase = '$this->{dbase}'?" unless $this->{dbase};
    $this->{dbh} = $this->{dbase} if ref($this->{dbase}) eq "DBI::db";
    $this->{trace} = shift;

    $this->{dbh}->trace( $this->{trace} ) if $this->{dbh};

    return $this;
}
# }}}
# do {{{
sub do {
    my $this = shift; return unless @_;

    my $r; eval { $r = $this->ready(shift)->execute(@_) }; croak $this->errstr if $@;
    return $r;
}
# }}}
# light_lock {{{
sub light_lock {
    my $this   = shift; return unless @_;
    my $tolock = join(", ", map("$_ read", @_));

    $this->handle->do("lock tables $tolock");
}
# }}}
# lock {{{
sub lock {
    my $this   = shift; return unless @_;
    my $tolock = join(", ", map("$_ write", @_));

    $this->handle->do("lock tables $tolock");
}
# }}}
# unlock {{{
sub unlock {
    my $this = shift;

    $this->handle->do("unlock tables");
}
# }}}
# ready {{{
sub ready {
    my $this = shift;

    return new MySQL::Easy::sth( $this, @_ );
}
# }}}
# firstcol {{{
sub firstcol {
    my $this = shift;
    my $query = shift;

    return $this->handle->selectcol_arrayref($query, undef, @_);
}
# }}}
# thread_id {{{
sub thread_id {
    my $this = shift;

    return $this->handle->{mysql_thread_id};
}
# }}}
# last_insert_id {{{
sub last_insert_id {
    my $this = shift;

    # return $this->firstcol("select last_insert_id()")->[0];
    return $this->handle->{mysql_insertid};
}
# }}}
# errstr (needs to be here, called from AUTOLOAD) {{{
sub errstr {
    my $this = shift;

    return $this->handle->errstr;
}
# }}}
# trace (needs to be here, called from AUTOLOAD) {{{
sub trace {
    my $this = shift;

    $this->handle->trace( @_ );
}
# }}}
# DESTROY {{{
sub DESTROY {
    my $this = shift;

    $this->{dbh}->disconnect if $this->{dbh};
}
# }}}
# handle {{{
sub handle {
    my $this = shift;

    return $this->{dbh} if defined($this->{dbh}) and $this->{dbh}->ping;
    # warn "WARNING: MySQL::Easy is trying to reconnect (if possible)" if defined $this->{dbh};

    ($this->{user}, $this->{pass}) = $this->unp unless $this->{user} and $this->{pass};

    $this->{host}  = "localhost" unless $this->{host};
    $this->{port}  =      "3306" unless $this->{port};
    $this->{dbase} =      "test" unless $this->{dbase};
    $this->{trace} =           0 unless $this->{trace};

    $this->{dbh} =
    DBI->connect("DBI:mysql:$this->{dbase}:host=$this->{host}:port=$this->{port}:mysql_compression=1;mysql_auto_reconnect=1", 
        $this->{user}, $this->{pass});

    croak "failed to generate connection: " . DBI->errstr unless $this->{dbh};

    $this->{dbh}->trace($this->{trace});

    return $this->{dbh};
}
# }}}
# unp {{{
sub unp {
    my $this = shift;

    my ($user, $pass);
    my $mycnf = "$ENV{HOME}/.my.cnf";

    open PASS, $mycnf or die "erorr reading $mycnf to find (otherwise unspecified) username and password: $!";

    my $l;
    while($l = <PASS>) {
        $user = $1 if $l =~ m/user\s*=\s*(.+)/;
        $pass = $1 if $l =~ m/password\s*=\s*(.+)/;

        last if($user and $pass);
    }

    close PASS;

    return ($user, $pass);
}
# }}}
# set_host set_user set_pass {{{
sub set_host { 
    my $this = shift;

    $this->{host} = shift;
}

sub set_port { 
    my $this = shift;

    $this->{port} = shift;
}

sub set_user { 
    my $this = shift;
    
    $this->{user} = shift;
}

sub set_pass { 
    my $this = shift;
    
    $this->{pass} = shift;
}
# }}}
# bind_execute {{{
sub bind_execute {
    my $this = shift;
    my $sql  = shift;

    my $sth = $this->ready($sql);

    $sth->execute            or return undef;
    $sth->bind_columns( @_ ) or return undef;

    return $sth;
}
# }}}

__END__
# Below is stub documentation for your module. You better edit it!

=head1 NAME

MySQL::Easy - Perl extension to handle various mundane DBI session related things specific to mysql.

=head1 SYNOPSIS

  use MySQL::Easy;

  my $trace = 0; # the trace arg is optional
  my $dbo = new MySQL::Easy("stocks", $trace);

  #  This is NEW (and totally untested):
  #  $dbo = new MySQL::Easy($existing_DBI_dbh, $trace);

  my $symbols = $dbo->firstcol(
      qq( select symbol from ohlcv where symbol != ?),
      "msft"
  );

  my $q = $dbo->ready("select * from ohlcv where symbol=?");
  for my $s (@$symbols) {
      my @a;

      $q->execute($s) or die $dbo->errstr;
      # Although the regular DBI will reconnect to the server when it loses
      # it's connection, any queries you had previously prepared will fail
      # until it reconnects.  easy::mysql handles all that for you so this
      # execute() will always work, if it's syntactically correct, even if
      # your server connection was lost since the last ready();

      print "@a" while @a = fetchrow_array $q;
  }

=head1 DESCRIPTION

   I don't remember how I used to live without this...
   I do like the way DBI and DBD work, but I wanted something
   _slightly_ prettier... _slightly_ handier.

   Here's the functions MySQL::Easy provides:

   $dbo = new MySQL::Easy( $db_name, $trace );
       # $db_name is the name of the database you're connecting to...
       # If you don't pick anything, it'll pick "test" for you.
       # $trace is a 1 or false, ... it's the DBI->trace() ...

   $dbo->do("sql statement bind=? bind=?", $bind1, $bind2);
       # this immediately executes the sql with the bind vars
       # given.  You can pas in a statement handle
       # instead of the string... this is faster if you're going
       # to use the sql over and over.  Returns a t/f like you'd
       # expect.  (i.e. $dbo->do("stuff") or die $dbo->errstr);

   $dbo->lock("table1", "table2", "table3");
       # MySQL::Easy uses only write locks.  Those are the ones
       # where nobody can read or write to the table except the
       # locking thread.  If you need a read lock, let Jet know.
       # Most probably though, if you're using this, it's a
       # smaller app, and it doesn't matter anyway.
   $dbo->unlock;

   $sth = $dbo->ready("Sql Sql Sql=? and Sql=?");
       # returns a DBI statement handle...
       # $sth->execute($bindvar); $sth->fetchrow_hashref; etc...

   $arr = $dbo->firstcol("select col from tab where x=? and y=?", $x, $y)
       # returns an arrayref of values for the sql.
       # You know, print "val: $_\n" for @$arr;
       # very handy...

   $id = $dbo->last_insert_id;
       # self explainatory?

   $dbo->trace(1); $dbo->do("sql"); $dbo->trace(0);
       # turns the DBI trace on and off.

   $dbo->errstr
       # returns an error string for the last error on the
       # thread...  Same as a $sth->errstr.  It's actually
       # described in DBI

   $dbo->check_warnings
       # I'll just give this example:
       $dbo->do("create temporary table cool( field enum('test1', 'test2') not null )");
       $dbo->do("insert into cool set field='test3'");
       $dbo->check_warnings 
           or die "SQL WARNING: $@\twhile inserting test field\n\t";

   $dbo->set_host($h); $dbo->set_port($p); 
   $dbo->set_user($U); $dbo->set_pass($p);
       # The first time you do a do/ready/firstcol/etc,
       # MySQL::Easy connects to the database.  You may use these
       # set functions to override values found in your ~/.my.cnf
       # for user and pass.  MySQL::Easy reads _only_ the user
       # and pass from that file.  The host name will default to
       # "localhost" unless explicitly set.  Also, it will die on
       # a fatal error if the user or pass is false and the
       # ~/.my.cnf cannot be opened.

   my $table;
   my $sth = $dbo->bind_execute("show tables", \( $table ) );
       # This was Josh's idea... And a good one.

   die $dbo->errstr unless $sth;  
       # bind_execute returns undef if either the bind
       # or execute phases fail.

   print "$table\n" while fetch $sth;

   # Anything from the DBI.pm manpage will work with the
   # $dbo (thanks to an AUTOLOAD function).

=head1 AUTHOR

Jettero Heller <japh@voltar-confed.org>

Jet is using this software in his own projects...
If you find bugs, please please please let him know. :)

Actually, let him know if you find it handy at all.
Half the fun of releasing this stuff is knowing 
that people use it.

=head1 THANKS

For bugs and ideas: Josh Rabinowitz <joshr-cpan@joshr.com>

=head1 COPYRIGHT

GPL!  I included a gpl.txt for your reading enjoyment.

Though, additionally, I will say that I'll be tickled if you were to
include this package in any commercial endeavor.  Also, any thoughts to
the effect that using this module will somehow make your commercial
package GPL should be washed away.

I hereby release you from any such silly conditions.

This package and any modifications you make to it must remain GPL.  Any
programs you (or your company) write shall remain yours (and under
whatever copyright you choose) even if you use this package's intended
and/or exported interfaces in them.

=head1 SEE ALSO

perl(1), DBI(3)

=cut
